Fix KeyError in s3fs _prune_deleted_files function

GRomR1 opened the pull request:
Fixes issue #68335 where fileserver.update fails with KeyError: 'Key' when using S3 buckets in multiple environments per bucket mode.

## Problem
The issue occurs in the `_prune_deleted_files` function where the code assumes a different data structure than what is actually provided. When using multiple environments per bucket mode, the metadata structure is nested differently than expected.

## Solution
The fix properly handles the nested bucket structure by iterating through buckets and objects within each bucket to extract the 'Key' field correctly.

## Changes Made

### 1. Core Fix
- Modified `_prune_deleted_files` function in `salt/fileserver/s3fs.py`
- Fixed KeyError when processing metadata in multiple environments per bucket mode
- Added proper iteration through bucket structure to handle nested metadata

### 2. Code Improvements
- Applied reviewer feedback from @bdrx312 to make code more Pythonic
- Changed from `meta.keys()` to `meta.values()` for cleaner iteration
- Improved code readability while maintaining functionality

### 3. Test Coverage
- Added comprehensive tests for both s3fs configuration modes:
  - `test_prune_deleted_files_multiple_envs_per_bucket` - Tests the fix for multiple environments per bucket mode
  - `test_prune_deleted_files_single_env_per_bucket` - Tests single environment per bucket mode to prevent regression
- Tests verify:
  - Files are properly cached initially
  - Deleted files are removed from cache when calling `_prune_deleted_files`
  - Remaining files continue to exist in cache
  - The nested metadata structure is handled correctly in multiple env mode

## Testing
This fix resolves the KeyError that was preventing fileserver.update from working with S3 buckets in multiple environments per bucket mode. The added tests ensure the fix works correctly and prevent future regressions.

## Related Issues
Fixes #68335

Signed-off-by: GRomR1 <gromrx1@gmail.com>

GRomR1 committed:
Fix KeyError in s3fs _prune_deleted_files function

Fixes issue #68335 where fileserver.update fails with KeyError: 'Key'
when using S3 buckets in multiple environments per bucket mode.

The issue occurs in _prune_deleted_files function where the code assumes
a different data structure than what is actually provided. The fix properly
handles the nested bucket structure by iterating through buckets and objects
within each bucket to extract the 'Key' field correctly.

Signed-off-by: GRomR1 <gromrx1@gmail.com>

GRomR1 committed:
Add changelog entry for issue #68335 fix

Signed-off-by: GRomR1 <gromrx1@gmail.com>

bdrx312 commented on salt/fileserver/s3fs.py:
```diff
@@ -591,7 +591,9 @@ def _prune_deleted_files(metadata):
             for meta in env_data:
-                cached_files.add(meta["Key"])
+                for bucket in meta.keys():
+                    for obj in meta[bucket]:
```
To iterate over items (keyword, value pairs) in a dict, use `.items()`, but this isn't using the key so just use `.values()`

```python
for objects in meta.values():
    for obj in objects:
```

twangboy reviewed (changes requested):
Please write a test for this as well

GRomR1 committed:
Improve code readability in s3fs _prune_deleted_files function

Use .values() instead of .keys() for better Pythonic code style.
This addresses the review comment from bdrx312.

Signed-off-by: GRomR1 <gromrx1@gmail.com>

GRomR1 committed:
Add tests for s3fs _prune_deleted_files function

Add comprehensive tests for both single environment per bucket and
multiple environments per bucket modes to ensure the KeyError fix works
correctly in both scenarios.

Tests cover:
- File deletion from cache when files are removed from S3
- Proper handling of nested metadata structure in multiple env mode
- Cache cleanup for both bucket configuration modes

Signed-off-by: GRomR1 <gromrx1@gmail.com>

dwoz commented:
@GRomR1 Is this bug present on 3006.x too?

Commit:
Fix lint and assertions in s3fs _prune_deleted_files tests

Use the existing configure_loader_modules fixture instead of direct
assignment to s3fs.__opts__ / s3fs.__utils__ (which tripped pylint's
unmocked-patch-dunder rule). Override s3.buckets via patch.dict for the
multi-env-per-bucket scenario.

Correct the cached-file path assertions: in multi-env-per-bucket mode
the env prefix is part of the S3 key, so _get_cached_file_name() returns
<cachedir>/<env>/<bucket>/<env>/<filename>, not
<cachedir>/<env>/<bucket>/<filename>.

The multi-env test now asserts the property the PR actually fixes -
_prune_deleted_files returning without raising KeyError on the metadata
shape produced by _init() in multi-env mode.

Co-authored-by: GRomR1 <gromrx1@gmail.com>

twangboy commented on salt/fileserver/s3fs.py:
```diff
@@ -591,7 +591,9 @@ def _prune_deleted_files(metadata):
-                cached_files.add(meta["Key"])
+                for objects in meta.values():
+                    for obj in objects:
+                        cached_files.add(obj["Key"])
```
"While this correctly fixes the KeyError for setups using multiple environments per bucket, changing the iteration logic this way introduces a regression for standard, flat env_data structures.

If a user has a standard configuration where meta is a flat dictionary containing the "Key" directly, meta.values() will extract string primitives instead of lists or sub-dictionaries. When the script hits the nested for obj in objects: loop, it will iterate over the characters of that string, leading to a TypeError or another KeyError when trying to look up obj["Key"].

We should handle this defensively to support both data structures:
```python
for meta in env_data:
    if "Key" in meta:
        cached_files.add(meta["Key"])
    else:
        # Handle the nested structure for multi-environment buckets
        for objects in meta.values():
            if isinstance(objects, list):
                for obj in objects:
                    if isinstance(obj, dict) and "Key" in obj:
                        cached_files.add(obj["Key"])
```

twangboy commented:
Please rebase and address conflicts and comments.
